### ISSUE TITLE
Benchmarks vs browser cache

### ISSUE THREAD

**Issue by re-thc** [i18next/i18next-localstorage-backend#21]
Hi, when is a good time to use this backend vs relying on the browser cache? Localstorage isn't particularly fast. Has there been a benchmark that shows this is faster? Often re-fetching it is even faster than localstorage retrieval in cases like mobile. Thanks.

**Comment by jamuhl** [comment]
good question...which I can't really give an answer as I personally never had the need for caching in localStorage -> but there are devs using this...so must make sense in some scenarios

**Comment by darshak369** [comment]
@jamuhl I am using browser cache (local storage) in my product to store the all translations JSON file. I am facing issue while implementing to this package to my product.

This is i18n.js file referring from https://www.i18next.com/how-to/caching#browser-caching-with-local-storage
```js
i18n
  // pass the i18n instance to react-i18next.
  .use(Backend)
  .use(initReactI18next)
  // init i18next
  // for all options read: https://www.i18next.com/overview/configuration-options
  .init({
    backend: {
      backends: [
        LocalStorageBackend, // primary backend
        HttpApi, // fallback backend
      ],
      backendOptions: [
        {
          /* options for primary backend */
          expirationTime: 7 * 24 * 60 * 60 * 1000, // 7 days
        },
        {
          /* options for secondary backend */
        },
      ],
    },
    lng: 'en-US',
    fallbackLng: false,
    ns: [],
    // debug: true,
    load: 'currentOnly',
    keySeparator: '.',

    interpolation: {
      escapeValue: false, // react already safes from xss,
    },
  });
```

I am using `useTranslation` hook for translating the text to different language.

Example.js 
```js
import React, { useEffect, useRef } from 'react';
import { useTranslation } from 'react-i18next';

function Filters() {
  const { t: translateReport } = useTranslation('reports');
  return(
        <Button>
          {translateReport('filters.buttons.clearAll')}
        </Button>,
        <Button>
          {translateReport('filters.buttons.apply')}
        </Button>,
  );
}
```

here is the `report.json` translation file
```js
{
  "filters": {
    "title": "Filters",
    "buttons": {
      "clearAll": "Clear All Filters",
      "apply": "Apply"
    },
    "submittedAt": "Submitted at",
 }
}

``` 

Whenever I change something on JSON file Its not reflecting to local storage. I have to clear the cache from local storage in browser and then again have reload whole react application then the changes got reflected in JSON file.

Currently I am thinking of two approach 
1. When JSON file modify needs to clear the local storage.
2. When any JS file modify in code then will clear the local storage.

It will be very helpful if you share some insight here.

**Comment by jamuhl** [comment]
Your two approaches have two issues. The backend runs in the browser and has no idea about if you modified a JSON.

use expiration time or versions to invalidate the cache: https://github.com/i18next/i18next-localstorage-backend#cache-backend-options

**Comment by jamuhl** [comment]
And during development don't use it at all

**Comment by darshak369** [comment]
Thanks for your reply @jamuhl 

I have try to use versions one thing I found is that when I update the JSON it and make use of versions like 
`versions: { en: 'v1.2' }`
It works fine yet when I made change second time on JSON it not working because we need to update the version whenever we updates the JSON file.

is that what I supposed to do ?

**Comment by darshak369** [comment]
Thanks @jamuhl 
I have misunderstand it actually now I got your point after reading below issue - 
[https://github.com/i18next/i18next-localstorage-backend/issues/3](url) 

Every release we have to change the version of JSON which will update the JSON file in production
correct me if I am wrong !

**Comment by jamuhl** [comment]
correct.

**Comment by KrisKrasovska** [comment]
@jamuhl Hello! Help me please, what am I doing wrong? Why is it that when I change something in a translation file, it is not immediately changed locally in localStorage, and if updates are sent to production, then the added transitions are not displayed there immediately, only after re-login, but I would like it to work automatically
file generate-i18n-hash.mjs
import crypto from "crypto";
import fs from "fs";
import {dirname, resolve} from "path";
import {fileURLToPath} from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

function getAllFiles(dir) {
    const entries = fs.readdirSync(dir, {withFileTypes: true});
    return entries.flatMap(entry => {
        const fullPath = resolve(dir, entry.name);
        return entry.isDirectory() ? getAllFiles(fullPath) : [fullPath];
    });
}

function hashFiles(files) {
    const hash = crypto.createHash("sha256");
    files.forEach(file => {
        const content = fs.readFileSync(file);
        hash.update(content);
    });
    return hash.digest("hex");
}

const localesPath = resolve(__dirname, "../public/locales");
const versionFilePath = resolve(localesPath, "i18n-version.json");

const files = getAllFiles(localesPath).filter(file => !file.endsWith("i18n-version.json"));
const hash = hashFiles(files);
console.log("hash:", hash);
const versionJson = {
    i18nHash: hash,
};

fs.writeFileSync(versionFilePath, JSON.stringify(versionJson, null, 2), "utf-8");

package.json
...
 "scripts": {
    "generate-i18n-hash": "node scripts/generate-i18n-hash.mjs",
    "dev": "yarn start",
    "dev:android": "npx cap run android -l --external",
    "build": "yarn generate-i18n-hash && react-scripts build",
    "build:android": "yarn build && npx cap sync && yarn copy:adjust && npx cap run android --no-sync",
    "start": "yarn generate-i18n-hash && react-scripts start",
...

i18n.ts
import {initReactI18next} from "react-i18next";
import {toast} from "react-toastify";

import i18n from "i18next";
import LanguageDetector from "i18next-browser-languagedetector";
import ChainedBackend from "i18next-chained-backend";
import HttpBackend from "i18next-http-backend";
import LocalStorageBackend from "i18next-localstorage-backend";
import {isDev} from "./utils/evironment";
// don't want to use this?
// have a look at the Quick start guide
// for passing in lng and translations on init

i18n
    .use(ChainedBackend)
    // detect user language
    // learn more: https://github.com/i18next/i18next-browser-languageDetector
    .use(LanguageDetector)
    // pass the i18n instance to react-i18next.
    .use(initReactI18next);

const initI18n = async() => {
    const i18nHashState = {i18nHash: ""};

    try {
        const prefix = isDev() ? "" : `${process.env.REACT_APP_ROUTE_PREFIX}`;
        const versionUrl = `${prefix}/locales/i18n-version.json`;

        try {
            const response = await fetch(versionUrl);

            const contentType = response.headers.get("Content-Type") || "";

            if (!response.ok || !contentType.includes("application/json")) {
                throw new Error(`Некорректный ответ при загрузке i18n-version.json. Тип контента: ${contentType}`);
            }

            const json = await response.json();

            if (!("i18nHash" in json)) {
                throw new Error("Поле 'i18nHash' не найдено.");
            }

            if (!json.i18nHash) {
                throw new Error("Пустое значение ключа 'i18nHash'");
            }

            i18nHashState.i18nHash = json.i18nHash;
        } catch (error) {
            toast.error(`Ошибка при загрузке i18n-version.json: ${error}`);
            // eslint-disable-next-line no-console
            isDev() && console.error("Ошибка при загрузке i18n-version.json:", error);
        }
    } finally {
        // init i18next
        // for all options read: https://www.i18next.com/overview/configuration-options
        await i18n.init({
            fallbackLng: "ru",
            supportedLngs: ["ru", "en"],
            backend: {
                backends: typeof window !== "undefined" ? [LocalStorageBackend, HttpBackend] : [],
                backendOptions: [
                    {
                        expirationTime: 60 * 60 * 1000, // 1 hour
                        versions: {ru: i18nHashState.i18nHash, en: i18nHashState.i18nHash},
                    },
                    {
                        loadPath: `${process.env.PUBLIC_URL}/locales/{{lng}}/{{ns}}.json`,
                    },
                ],
            },

            ns: ["dashboard"],
            defaultNS: "dashboard",

            debug: true,

            interpolation: {
                escapeValue: false, // not needed for react as it escapes by default
            },
            i18nFormat: {
                localeData: "ru",
            },
        });
    }
};

void initI18n();

export default i18n;
index.js
// импорт sentry должен быть выше остальных
// eslint-disable-next-line simple-import-sort/imports
import "./sentry";
import React from "react";
import ReactDOM from "react-dom/client";
import {Provider} from "react-redux";
import {BrowserRouter as Router} from "react-router-dom";
import * as Sentry from "@sentry/react";

import "./i18n"; 
import {makeStore} from "./store";

import App from "./App";

import "react-toastify/dist/ReactToastify.css";
import "rsuite/dist/rsuite.min.css";
import "./styles/globals.scss";

export const store = makeStore();

const root = ReactDOM.createRoot(document.getElementById("root"));

root.render(
    <React.StrictMode>
        <Sentry.ErrorBoundary fallback={<p>Что-то пошло не так</p>}>
            <Provider store={store}>
                <Router basename={process.env.PUBLIC_URL}>
                    <App/>
                </Router>
            </Provider>
        </Sentry.ErrorBoundary>
    </React.StrictMode>
);

**Comment by jamuhl** [comment]
that really depends on your server setup how it caches (sets ttl)

you might try adding some uniqueness to your loadPath like: `${process.env.PUBLIC_URL}/locales/{{lng}}/{{ns}}.json?refresh=${Date.now())` - might help - but no guarantee

**Comment by KrisKrasovska** [comment]
> that really depends on your server setup how it caches (sets ttl)
> 
> you might try adding some uniqueness to your loadPath like: `${process.env.PUBLIC_URL}/locales/{{lng}}/{{ns}}.json?refresh=${Date.now())` - might help - but no guarantee

@jamuhl Thank you very much for your answers! I hope it helps me
